metagraph/src/annotation/binary_matrix/row_sparse_disk/row_sparse_disk.cpp

```diff
@@ -0,0 +1,139 @@
+        rows[i] = get_row(row_ids[i]);
+    }
+
+    return rows;
+}
+
+bool RowSparseDisk::load(std::istream &f) {
+    auto _f = dynamic_cast<mtg::common::IfstreamWithNameAndOffset *>(&f);
+    assert(_f);
+    try {
+        num_columns_ = load_number(f);
+
+        auto boundary_start = load_number(f);
+
+        _f->set_offset(static_cast<uint64_t>(f.tellg()));
+
+        int_vector_buffer_params.filename = _f->get_name();
+        int_vector_buffer_params.offset = _f->get_offset();
+
+        f.seekg(-sizeof(std::streampos), ios_base::end);
+
+        f.seekg(boundary_start, ios_base::beg);
+        boundary_.load(f);
+
+        num_rows_ = boundary_.num_set_bits();
+
+    } catch (...) {
+        return false;
+    }
+
+    return true;
+}
+
+void RowSparseDisk::serialize(std::ostream & /*f*/) const {
+    throw std::runtime_error(
+            "RowSparseDisk::serialize not supproted, use static variant instead");
+}
+
+
+void RowSparseDisk::serialize(const std::function<void(binmat::BinaryMatrix::RowCallback)> &call_rows,
+                              const std::string &filename,
+                              uint64_t num_cols,
+                              uint64_t num_set_bits,
+                              uint64_t num_rows) {
+    std::ofstream outstream(filename,
+                            std::ios::binary | std::ios::ate
+                                    | std::ios::in); // std::ios_base::in needed ...
+
+    if (!outstream.good())
+        throw std::ofstream::failure("Cannot write to file " + filename);
+
+    serialize_number(outstream, num_cols);
+    auto boundary_start_pos = outstream.tellp();
+    // write "empty" boundary start
+    std::streampos boundary_start = 0;
+    serialize_number(outstream, boundary_start);
+
+    const uint64_t iv_offs = outstream.tellp();
+    outstream.close();
+
+    auto outbuf = sdsl::int_vector_buffer<>(filename, std::ios::out | std::ios::binary,
+                                            1024 * 1024, sdsl::bits::hi(num_cols) + 1,
+                                            false, iv_offs);
+    // call_bits should be called for each argument that should be set (row boundaries)
+    auto call_bits = [&](const std::function<void(uint64_t)> &call_bit) {
+        uint64_t t = 0;
+        call_rows([&](const auto &row) {
+            uint64_t last = 0;
+            for (auto val : row) {
+                outbuf.push_back(val - last);
+                last = val;
+            }
+            call_bit((t += row.size() + 1) - 1);
+        });
+        outbuf.close();
+    };
+
+    bit_vector_small boundary(call_bits, num_rows + num_set_bits, num_rows);
+
+    outstream.open(filename, std::ios::in | std::ios::out | std::ios::binary | std::ios::ate);
+    boundary_start = outstream.tellp();
+    boundary.serialize(outstream);
+    outstream.seekp(boundary_start_pos, ios_base::beg);
+    outstream.write(reinterpret_cast<char *>(&boundary_start), sizeof(std::streampos));
```

Comment:
Since you have to load `boundary` anyway, why not serialize them before the other vector? Then you won't need to do another seek in load. In fact, you won't need to do any seek in load...

Reply:
Yeah, but the boundary is built in memory while writing rows data to disk via `int_vector_buffer`. If my understanding is correct it would require iterating rows twice to first build the boundary, store it and then iterate again to store rows' content. 
I assumed we don't want to pay the cost of the second iteration.
One option would be to leave the space for a boundary before rows content, yet I think we don't know (in advance) how many bytes are needed to store the boundary (to prepare an appropriate "hole" in the file), or is there a way to determine it?
Maybe there is another way that I just can't see. Any ideas?

Reply:
Yeah, I see the point. Would be impossible without a temp file. I guess it's better to keep it as is then. Though would be good to have a clarifying comment

Reply:
I have added some comments to the code, I hope it clarifies things